client/src/router/Router.jsx

```diff
@@ -0,0 +1,102 @@
+import { useState } from "react";
+import { useEffect } from "react";
+import { navigate } from "./navigate";
+
+/**
+ * @param {Object} props
+ * @param {Object.<string, React.ComponentType>[]} props.routes
+ * @returns {React.JSX.Element}
+ */
+function Router({ routes }) {
+  const [currentPath, setCurrentPath] = useState(window.location.pathname);
+  const [Component, setComponent] = useState(<NotFound />);
+
+  useEffect(() => {
+    const onLocationChange = () => {
+      setCurrentPath(window.location.pathname);
+    };
+
+    window.addEventListener("popstate", onLocationChange);
+
+    return () => {
+      window.removeEventListener("popstate", onLocationChange);
+    };
+  }, []);
+
+  useEffect(() => {
+    for (const routeObj of routes) {
+      const [path, Comp] = Object.entries(routeObj)[0];
+      const isMatch = matchParams(path, currentPath);
+      if (isMatch !== null) {
+        setComponent(<Comp params={isMatch} />);
+        return;
+      }
+    }
+    setComponent(<NotFound />);
+  }, [currentPath, routes]);
+
+  return Component;
+}
+
+/**
+ * Link compnent for SPA style navigation
+ * @param {Object} props
+ * @param {string} props.to
+ * @param {React.ReactNode} props.children
+ * @returns {JSX.Element}
+ */
+function Link({ to, href, children, ...props }) {
+  const handleClick = (e) => {
+    e.preventDefault();
+    navigate(to || href);
+  };
+
+  return (
+    <a href={to} onClick={handleClick} {...props}>
+      {children}
+    </a>
+  );
+}
+
+function NotFound() {
+  return (
+    <div>
+      Not found <Link to="/">return Home</Link>
+    </div>
+  );
+}
+
+/**
+ *
+ * @param {string} routePath
+ * @param {string} currentPath
+ * @returns {Object<string, string> | null} - a key value map or the seach params.
+ */
+function matchParams(routePath, currentPath) {
```

Comment:
```suggestion
function matchParams(routePattern, browserPath) {
```
I got confused on the difference between the paths, this is more accurate, right?